NFSv3 and NFSv4 fixes and improvements

thexai opened the pull request:
## Description
NFSv3 and NFSv4 fixes and improvements:

- Fixes session expiration due insufficient keep alive interval (NFSv4 only).
- Fixes attempt to reuse cached but expired NFS contexts (NFSv4 only).
- Fixes crash when read Blu-Ray BDMV structures and some optional files not exist (NFSv3 & NFSv4).
- Use 128K read/write chunk size in NFSv4  (NFSv4 servers do not enforce or limit to any specific value).
- Fixes https://github.com/xbmc/xbmc/issues/22862
- Fixes https://github.com/xbmc/xbmc/pull/22714#issuecomment-1439719284

## Motivation and context
Due to the big differences between NFSv3 and NFSv4, certain changes are necessary, especially everything related to time outs and keep alive:

> NFSv3 is stateless. That effectively means that the NFS server (ONTAP) doesn't keep track of which filesystems are mounted, by whom, or which locks are truly in place. Tracking NFS client state is not part of the NFSv3 standard.
> 
> In contrast, NFSv4 is stateful. The NFSv4 server tracks which clients are using which filesystems, which files exist, which files and/or regions of files are locked, etc. This means there needs to be regular communication between an NFSv4 server to keep the state data current.

https://community.netapp.com/t5/Tech-ONTAP-Blogs/NFSv3-and-NFSv4-What-s-the-difference/ba-p/441316

> The client queries the server's "lease_time" attribute. (Default 90s on Linux.)
> Normal filesystem activity starts. The client ensures that it sends one rpc (may be some sort of no-op) at least every lease_time, to reassure the server that it's still active.

> NFSv3 also has the notion of grace periods and reclaim flags on lock requests. The main differences are:
> there's no lease time, and clients aren't required to poll the server regularly.

https://linux-nfs.org/wiki/index.php/NFS_lock_recovery_notes

## How has this been tested?
Runtime Windows x64 and Android (Shield). 
As NFSv4 server used Synology NAS (DSM 7.1.1) Indeed, it has been proven that it has 90s "lease_time".
Tested NFSv3 and NFSv4, tested and reproduced all issues mentioned.

## What is the effect on users?
- Fixes random `NFS4ERR_EXPIRED` errors at open files any time.
- Fixes video playback stalls and closes after pause due insufficient keep alive of NFSv4 session.
- Fixes crash opening Blu-Ray folders (BDMV) if some optional files not present (very common).
- Fixes slow NFSv4 read when not using file cache due very small read chunk size.


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai committed:
NFSFile: Only retry nfs_open when error is NFS4ERR_EXPIRED

Retrying with other error codes and force deinit connection may cause
crash, specifically when a file does not exist.

Anyway, NFS4ERR_EXPIRED should not occur now that contexts expiration and keep alives are
handled correctly.

thexai committed:
NFSFile: adapt timeouts for compatibility with NFSv4

thexai committed:
NFSFile: Log errors on keep alive reads

In case an NFSv4 server are configured with low "lese_time" i.e: 30s
the session will expire before receiving the first keep alive and we
will realize it by nfs_read errors here

graysky2 commented:
I build the latest commit form Nexus branch with this PR and found that all NFS paths are broken.  Can you confirm?

EDIT: my system is using a shared mysqld data base if that matters.  I am finding lines in the log like this:
```
2023-03-03 15:01:21.611 T:718      info <general>: Running database version Addons33
2023-03-03 15:01:21.611 T:718      info <general>: Running database version ViewModes6
2023-03-03 15:01:21.612 T:718      info <general>: Running database version Textures13
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic82 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic81 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic80 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic79 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic78 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic77 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.615 T:718     error <general>: Unable to open database: MyMusic76 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic75 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic74 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic73 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic72 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic71 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic70 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic69 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic68 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic67 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic66 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic65 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.616 T:718     error <general>: Unable to open database: MyMusic64 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic63 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic62 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic61 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic60 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic59 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic58 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic57 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic56 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic55 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic54 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.617 T:718     error <general>: Unable to open database: MyMusic53 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic52 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic51 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic50 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic49 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic48 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic47 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic46 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic45 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic44 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.618 T:718     error <general>: Unable to open database: MyMusic43 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic42 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic41 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic40 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic39 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic38 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic37 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic36 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic35 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic34 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic33 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.619 T:718     error <general>: Unable to open database: MyMusic32 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.620 T:718     error <general>: Unable to open database: MyMusic82 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.620 T:718     error <general>: Unable to create new database
2023-03-03 15:01:21.621 T:718     error <general>: Unable to open database: MyVideos121 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.621 T:718     error <general>: Unable to open database: MyVideos120 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.621 T:718     error <general>: Unable to open database: MyVideos119 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.621 T:718     error <general>: Unable to open database: MyVideos118 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.621 T:718     error <general>: Unable to open database: MyVideos117 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.621 T:718     error <general>: Unable to open database: MyVideos116 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos115 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos114 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos113 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos112 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos111 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos110 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos109 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos108 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos107 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos106 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos105 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.622 T:718     error <general>: Unable to open database: MyVideos104 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos103 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos102 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos101 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos100 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos99 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos98 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos97 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos96 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos95 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos94 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.623 T:718     error <general>: Unable to open database: MyVideos93 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos92 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos91 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos90 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos89 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos88 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos87 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos86 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos85 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos84 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos83 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos82 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.624 T:718     error <general>: Unable to open database: MyVideos81 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos80 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos79 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos78 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos77 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos76 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos75 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database: MyVideos121 [2002](Can't connect to server on '10.9.8.112' (101))
2023-03-03 15:01:21.625 T:718     error <general>: Unable to create new database
```

fritsch commented:
Sorry - this error has absolutely nothing to do with your mysql connection issue.

thexai commented:
`2023-03-03 15:01:21.625 T:718     error <general>: Unable to open database:`

This is totally unrelated: database/Mysql not uses NFS at all...

thexai committed:
NFSFile: use 128K read/write chunk size in NFSv4

NFSv4 not has maximun chunk size limitation at server side (used
TCP vs UDP in NFSv3, etc.) then 'nfs_get_readmax' returns zero or not
implemented.

Clients must set the chunk size they want to use.

graysky2 commented:
Very odd/I applied the latest version of this PR to the latest commit from Nexus and it is behaving as expected.  Not sure what triggered the errors I posted above.

thexai commented:
Note that last commit (128K chunks) is critical. 

Without it when FileCache is not used, and with some file types (as playing BDMV folders) never is used (even if buffermode is set to 1). VideoPlayer stalls with buffering all the time as 4K read chunks is not enougth for high bitrate streaming....

arnova commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -432,11 +444,13 @@ void CNfsConnection::keepAlive(const std::string& _exportPath, struct nfsfh* _pF
-  nfs_read(pContext, _pFileHandle, 32, buffer);
+  int bytes = nfs_read(pContext, _pFileHandle, 32, buffer);
+  if (bytes < 0)
+    CLog::LogF(LOGERROR, "nfs_read - Error ({}, {})", bytes, nfs_get_error(pContext));
```
Shouldn't we early return here to skip the SEEK_SET below?

arnova commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -585,34 +599,34 @@ bool CNFSFile::Open(const CURL& url)
+  int ret = nfs_open(m_pNfsContext, filename.c_str(), O_RDONLY, &m_pFileHandle);
 
-  if (!NfsOpen(url, filename))
+  if (ret == -11) // NFS4ERR_EXPIRED
```
Perhaps use a #define ?

arnova reviewed (commented):
Some minors but overall looking good.

lrusak commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -45,16 +45,16 @@ using namespace std::chrono_literals;
 
 namespace
 {
+// Default "lease_time" on most Linux NFSv4 servers are 90s.
```
Because this value can be changed is why I avoided adjusting the timeouts in the first place.

thexai replied:
But Linux default (nfs-kernel-server) is 90s. This is so....

6 minutes does NOT work. If you want to use 6 minutes because "avoid change here" then you need submit patch to Linux kernel

enen92 replied:
Wouldn't an advanced setting for this purpose be better (while keeping the sane defaults)?
I assume there's no easy way to get those values from the server

thexai replied:
> I assume there's no easy way to get those values from the server

In fact, the spec says that clients should get the lease_time from the server:

> The client queries the server's "lease_time" attribute. (Default 90s on Linux.)

https://linux-nfs.org/wiki/index.php/NFS_lock_recovery_notes

But we are limited to what `libnfs` offers and I haven't found a way to get this (note that this is "new" in NFSv4 and v4 support in libnfs is very limited and NFSv4.1 is unsupported).

Anyway 45s keep alive should work in all cases because common values are 90s (default), 120s or 60s.

30s or less is discouraged.

lrusak replied:
Maybe @sahlberg know more about the leasing period. It would be nice if we could just query it.

thexai replied:
I also think that it is necessary to clarify here that the keep alive that we have implemented (it was already in NFSv3) are not to keep the NFS server's session active (this concept simply did not exist in NFSv3) but rather to prevent the hard disk from stopping spin up and go into hibernation when the video is paused for several minutes.

It is true that now with NFSv4 the same keep alive works for both.

The keep alive for NFS session does not need to read data from the hard drive, it just sends a RENEW command. On the other hand, to keep the hard drive spinning, it is necessary to read something...

lrusak commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -45,16 +45,16 @@ using namespace std::chrono_literals;
-constexpr auto KEEP_ALIVE_TIMEOUT = 3min;
-
-constexpr auto IDLE_TIMEOUT = 3min;
+constexpr auto CONTEXT_TIMEOUT = 60s; // 2/3 parts of lease_time
```
Why do these other values need to be adjusted?

thexai replied:
Because context "expires" when time of inactivity is greater than server's "lease_time"

thexai replied:
Current solution is:

CONTEXT_TIMEOUT = 6 minutes
But expires in 90s then when someone opens a file after 90s nfs_open fails with NFS4ERR_EXPIRED but is retry with https://github.com/xbmc/xbmc/commit/9dc30e83c830d379c063eaf3a4a64a9fb57ca782


New solution is:

Discard expired contexts so `nfs_open` should never fail and should never retry...

lrusak commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -432,11 +444,13 @@ void CNfsConnection::keepAlive(const std::string& _exportPath, struct nfsfh* _pF
 
-  CLog::Log(LOGINFO, "NFS: sending keep alive after {} s.",
-            std::chrono::duration_cast<std::chrono::seconds>(KEEP_ALIVE_TIMEOUT).count());
+  CLog::LogF(LOGDEBUG, "sending keep alive after {}s.",
```
I imagine you changed this to debug because it's a bit spammy with the faster keep alive.

thexai replied:
True

thexai replied:
I want to prevent someone from opening a new issue "log spamming when video is paused"  and you assigning it to me. :)

lrusak commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -585,34 +599,34 @@ bool CNFSFile::Open(const CURL& url)
-    {
-      CLog::Log(LOGERROR, "CNFSFile::Open: Unable to open file: '{}' error: '{}'",
-                url.GetFileName(), nfs_get_error(m_pNfsContext));
+  if (ret != 0)
```
Are positive error codes allowed?

thexai replied:
only `ret == 0` is open successfully, then `ret != 0` is open FAILED which is precisely what we want to detect here

lrusak commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -45,16 +45,16 @@ using namespace std::chrono_literals;
 constexpr auto SETTING_NFS_VERSION = "nfs.version";
-
-} // namespace
+} // unnamed namespace
```
I'm surprised clang-format doesn't trip up on this.

thexai replied:
?

thexai replied:
https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#6-scoping

wsnipex commented on xbmc/filesystem/NFSFile.cpp:
```diff
@@ -327,10 +329,22 @@ bool CNfsConnection::Connect(const CURL& url, std::string &relativePath)
+    if (m_readChunkSize == 0)
+    {
+      CLog::Log(LOGDEBUG, "NFS Server did not return max read chunksize - Using 128K default");
+      m_readChunkSize = 128 * 1024; // 128K
```
I wonder if it's worth making this an advanced setting.

fritsch replied:
There is already chunksize as advancedsetting, that this works though you need to return 1 here,then FileCache will set it to 64K or to chunksize else. Sucks. Smb returns 128K always.

thexai replied:
@wsnipex 
It seems like a good idea to me but I will do in a follow up PR to keep this simple for backport.

128K has been chosen (and should be fine for all) because 128K is already used in NFSv3 and also when FileCahe is used with NFSv4.

This code force to use default 128K chunk size when source reports 0:

https://github.com/xbmc/xbmc/blob/56b88b2bf5d82fd2b2feb4df4b35f81e70ca1b89/xbmc/filesystem/FileCache.cpp#L138-L144

thexai replied:
File chunk size is a bit chaotic because is hardcoded in many places:

https://github.com/xbmc/xbmc/blob/56b88b2bf5d82fd2b2feb4df4b35f81e70ca1b89/xbmc/filesystem/File.cpp#L119

thexai replied:
> There is already chunksize as advancedsetting

This is chunk size for FileCahe only

We can create a new setting for NFS only (when is not used in combination with FileCache)

wsnipex replied:
yes, it's cache only, so won't work when caching is disabled.

adding this in a followup PR is fine for me.

wsnipex reviewed (approved):
looks good to me, but left a question for your consideration

lrusak commented:
Why was this rushed to be merged? There was still outstanding comments and no approvals 😞

thexai commented:
> and no approvals 

![approval](https://user-images.githubusercontent.com/58434170/222975138-3a3d025b-a800-44c4-9ee6-abcc75a5b119.png)
